In [25]:
import pandas as pd

In [26]:
data=pd.read_csv("C://Users//Ayush Jain//OneDrive//logistic//Logistic-regression_final//diabetes.csv")

In [27]:
data.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   
5            5      116             74              0        0  25.6   
6            3       78             50             32       88  31.0   
7           10      115              0              0        0  35.3   
8            2      197             70             45      543  30.5   
9            8      125             96              0        0   0.0   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
5                     0.201   30        0  
6                     0.248   26        1  
7                     0.134   29        0  
8                     0.158   53        1  
9                     0.232   54        1

checking the shape of the dataset to understand the features

In [28]:
data.shape

(768, 9)

In [7]:
data['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

In [8]:
data['Outcome'].value_counts(normalize=True).apply(lambda x:x*100)

0    65.104167
1    34.895833
Name: Outcome, dtype: float64

X here is our input dataset with input features needed for classification using SVM

In [9]:
y=data['Outcome']
y.shape

(768,)

In [10]:
X=data.copy()
X.pop('Outcome')
X.shape

(768, 8)

X and y have now been created as the input and output datasets respectively of the diabetes classification

In [11]:
X.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
2                     0.672   32  
3                     0.167   21  
4                     2.288   33

**checking the shape of the input and output datasets**

In [12]:
X.shape,y.shape

((768, 8), (768,))

There are 8 features that have been defined as input features and the corresponding labels of 768 records are defined as output features. 

Next, we will create the training and test patterns by keeping the 20% of data for testing purposes.

In [13]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=0)

The X_train and y_train are defined as input and output patterns for training respectively whereas the X_test and y_test are defined as input and output patterns for testing respectively. The model will be trained using the training pattern, tested using the X_test and the prediction results will be compared with y_test, the original class labels. 

Lets's check the shape of training and test data.

In [14]:
X_train.shape,X_test.shape

((614, 8), (154, 8))

In [15]:
print(X_train)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
603            7      150             78             29      126  35.2   
118            4       97             60             23        0  28.2   
247            0      165             90             33      680  52.3   
157            1      109             56             21      135  25.2   
468            8      120              0              0        0  30.0   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
192            7      159             66              0        0  30.4   
629            4       94             65             22        0  24.7   
559           11       85             74              0        0  30.1   
684            5      136             82              0        0   0.0   

     DiabetesPedigreeFunction  Age  
603                     0.692   54  
118                     0.443   22  


As there are values of some features which are higher than others, they will dominate the prediction results and hence need to be scaled. For this purpose, we will scale the training and test input sets. 

In [16]:
from sklearn.preprocessing import StandardScaler
s=StandardScaler()
X_train=s.fit_transform(X_train)
X_test=s.transform(X_test)

In [17]:
print(X_train)

[[ 0.90832902  0.91569367  0.44912368 ...  0.37852648  0.67740401
   1.69955804]
 [ 0.03644676 -0.75182191 -0.47230103 ... -0.50667229 -0.07049698
  -0.96569189]
 [-1.12606292  1.38763205  1.06340683 ...  2.54094063 -0.11855487
  -0.88240283]
 ...
 [ 0.03644676 -0.84620959 -0.21634972 ... -0.94927168 -0.95656442
  -1.04898095]
 [ 2.0708387  -1.12937261  0.24436264 ... -0.26640405 -0.50001442
   0.11706589]
 [ 0.32707418  0.47521786  0.65388473 ... -4.07275877  0.52121586
   2.94889395]]


**Step 3: Defining and training the SVM model**
In this step, we will define the support vector machine model for classification using the radial function (RBF) as the kernel function.

In [18]:
# Defining the SVM classification model
from sklearn.svm import SVC
SVM_classifier = SVC(kernel = 'linear', random_state = 0)

In [29]:
from sklearn.svm import SVR

In [19]:
# Training the SVM classifier
SVM_classifier.fit(X_train, y_train)

SVC(kernel='linear', random_state=0)

In [20]:
# Making predictions with the test data
y_pred = SVM_classifier.predict(X_test)

saving the results in a csv dataset

In [21]:
# Combining the predicted and actual values
output=pd.DataFrame(data={'Predicted Labels': y_pred, 'Actual Labels': y_test})
output.to_csv("output.csv")
print("Saved!")

Saved!


printing a confusion matrix to understand our final results

In [22]:
# Confusion matrix
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, y_pred))

[[98  9]
 [18 29]]


printing our final accuracy

In [23]:
# Accuracy score
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.8246753246753247

Finally, let's obtain the classification report for analyzing the performance of the classification model.

In [24]:
# Classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.84      0.92      0.88       107
           1       0.76      0.62      0.68        47

    accuracy                           0.82       154
   macro avg       0.80      0.77      0.78       154
weighted avg       0.82      0.82      0.82       154



Hence we have successfully used SVM for a diabetes dataset